src/library/scala/collection/immutable/List.scala

```diff
@@ -328,28 +332,45 @@ sealed abstract class List[+A] extends AbstractSeq[A]
   }
 
   final override def flatMap[B, That](f: A => GenTraversableOnce[B])(implicit bf: CanBuildFrom[List[A], B, That]): That = {
-    if (bf eq List.ReusableCBF) {
+    if (isLikeListReusableCBF(bf)) {
       if (this eq Nil) Nil.asInstanceOf[That] else {
         var rest = this
-        var found = false
-        var h: ::[B] = null
-        var t: ::[B] = null
+        class Appender extends Function1[B, Unit] {
+          var h: ::[B] = null
+          var t: ::[B] = null
+          override def apply(b: B): Unit = {
+              if (h eq null) {
+                h = new ::(b, Nil)
+                t = h
+              }
+              else {
+                val nx = new ::(b, Nil)
+                t.tl = nx
+                t = nx
+              }
+          }
+        }
+        var appender: Appender = null
         while (rest ne Nil) {
-          f(rest.head).seq.foreach{ b =>
-            if (!found) {
-              h = new ::(b, Nil)
-              t = h
-              found = true
-            }
-            else {
-              val nx = new ::(b, Nil)
-              t.tl = nx
-              t = nx
-            }
+          f(rest.head) match {
+            case c: LinearSeq[B] =>
```

Comment:
This kind of type test on interface types is really bad for Scala.js. This is doing way worse than good, compared to never optimizing for the empty case.

What is even the rationale, here? Avoiding the allocation of the `Appender`? Compared to the allocation of `f` in the first place, does this bring any advantage even on the JVM, not to mention the likely collections generated by `f`?

Reply:
@sjrd Not to mention that the growth in the bytecode size and additional branches will not improve anything either. One might argue that if "isEmpty" is not a cheap operation, that would be the responsibility of that data structure, not this.

Reply:
It's cheap in 2.13.x by design but not in 2.12.x, we've had regression's when we've assumed otherwise. :(

Reply:
I think we could dial this back to an optimization for `eq Nil` and live with allocating `Appender` in other cases when all elements map to empty sequences of other types.

Reply:
> What is even the rationale, here? Avoiding the allocation of the `Appender`? Compared to the allocation of `f` in the first place, does this bring any advantage even on the JVM, not to mention the likely collections generated by `f`?

There is no reason to suppose that f is allocated. It may be statically cached, and so may the results. It depends on the usage

This PR is all about reduction of allocations. Yes saving allocation is a real benefit on a JVM. GC is not free. If we cant come to a useful check on isEmpty then we may have to live with that

I understand that 2.13 has different expectations for isEmpty

Reply:
> One might argue that if "isEmpty" is not a cheap operation, that would be the responsibility of that data structure, not this.
The data structures that are offending are both in the scala libraries and in the community build. Some for example check `length != 0`

Reply:
FTR: here's the case where the extra call to `isEmpty` triggered a functional regression in an open source project: https://github.com/scala/scala-dev/issues/673#issuecomment-592814066

Reply:
@retronym @mkeskells What about `hasKnownSize && isEmpty`?

Reply:
The empty cases that I would expect to see would be mostly `Nil` and `None` (maybe streams `Empty` as well)
Maybe `case x if (x eq Nil) ||  (x eq None) ...`, which could just be without the pattern match?
```
val c = f(rest.head)
if ((c ne Nil) && (c ne None) && (c ne Empty) ) { …
```


            case c: LinearSeq[B] =>

Reply:
@mkeskells making correctly implemented solutions participate here seems preferable. What’s wrong with hasKnownSize && isEmpty?

Reply:
> @mkeskells making correctly implemented solutions participate here seems preferable. What’s wrong with hasKnownSize && isEmpty?

I am all in favour of correct implementation. (I presume that you mean `hasDefiniteSize`) I dont see **any** documented relationship between those two methods. Just because a collection has a finite size doesn't mean that determining if it is empty (or anything else about the size) is cheap for all collections in the standard library, let alone custom collections

It would be good to have some pattern that works with custom collections, but I cant see one

Reply:
reworked